(#3933) fixing improper texture allocation across cards

mattkae opened the pull request:
fixes #3933 

## What's new?
- `GLRenderingProvider::as_texture` now checks if we have a `ShmBuffer` and returns a texture for that specific provider instead of a single texture for all providers
- Implemented `ShmBuffer::texture_for_provider` which will return a `gl::Texture` either from a map cache or elsewhere
- Added `ShmBuffer::create_texture` which has a default implementation, but will be overridden for each of the subclasses. These subclasses mostly need to override the `bind()` method
- `ShmBuffer` no longer implements `gl::Texture`
- `ShmBuffer` no longer takes an `EGLContextExecutor` unless it is creating a texture
- Updated the associated unit tests

mattkae commented on src/platforms/common/server/kms-utils/kms_connector.cpp:
```diff
@@ -201,5 +203,11 @@ auto mgk::find_crtc_with_primary_plane(
         }
     }
 
+    if (plane_res.planes().begin() != plane_res.planes().end())
+    {
+        mir::log_error("Could not find primary plane for CRTC, using the first available one");
+        return std::make_pair(std::move(crtc), std::move(*plane_res.planes().begin()));
+    }
```
@RAOF @tarek-y-ismail 

I need to add this bit to make things work here, but this is obviously unrelated to this PR. I can move it to another PR, but is this even the right fix for this? I can't even start Mir on my machine with out this...

RAOF replied:
That is extremely weird? Also, that code hasn't changed in ages; why would you now need this? (This looks like a standup-tomorrow topic)

RAOF replied:
(This is actually https://github.com/canonical/mir/pull/3978)

RAOF commented:
I think this might be missing a few `as_texture()` implementations, and maybe there can be some simplifications? I need to spend some more time reviewing tomorrow.

mattkae commented:
> I think this might be missing a few `as_texture()` implementations, and maybe there can be some simplifications? I need to spend some more time reviewing tomorrow.

I think I just need to fix `renderer-generic-egl`, but the eglstreams one doesn't need saving since it is disappearing anyway, right?

RAOF reviewed (commented):
Hm. I *think* this can actually be dramatically simplified? I think you only need a single `ShmTexture` implementation; the differences are only how to upload the texture (and, a little annoyingly, needing to update the texture sometime for the `MemoryBacked` one).

If the `create_texture` implementations created a ShmTexture and also uploaded to that texture at the same time, you could have a single implementation (calling `on_consumed()` at `texture_for_provider()` time rather than `bind()` time [should be] is fine).

That leaves `mark_dirty`; this could be either just dropping the `provider_to_texture_map` when it's dirty (and re-allocating textures) or adding a `refresh` callback called in `texture_for_provider`?

RAOF reviewed (commented):
Looks better; I think this is pretty much ok.

Feel free to try to remove the `uploaded`/`mark_dirty()` from `ShmBufferTexture`, but if it doesn't go cleanly that's ok.

mattkae commented on src/platforms/common/server/shm_buffer.h:
```diff
@@ -53,24 +55,47 @@ class ShmBuffer :
+    private:
+        std::shared_ptr<EGLContextExecutor> egl_delegate;
+        GLuint tex_id_;
+        std::mutex uploaded_mutex;
+        bool uploaded = false;
```
Personally I like the idea of not reallocating `ShmBufferTexture` and just getting a single texture id and updating it. IMO, it is easier to have an `uploaded` per texture, as we know better when that particular texture has been uploaded and then we can handle that in `on_texture_accessed` for that texture.

RAOF reviewed (approved):
LGTM now.

I think the TICS findings are correct, though; this will reduce test coverage.

We could probably test the generic-egl-renderer by running some wlcs tests with that renderer? I'd need to check what lines we're missing in shm_buffer.cpp to see if that's something easily testable.

mattkae commented:
@RAOF 

So it seems like two things:

1. `buffer_allocator.cpp` is claiming that we reduced coverage, but TICS isn't even showing the lines that I added :thinking: In `mge::GLRenderingProvider::as_texture`, my new lines aren't shown
2. `shm_buffer.cpp` is mad about things that were preexisting, but that I moved. All new functionality appears to be tested :)

Saviq commented:
![image](https://github.com/user-attachments/assets/3cc3fa06-503e-4863-9639-95f0bad9622d)
